File: packages/core/src/finance/indicators.ts
/**
 * Technical-analysis indicators — pure array→array transforms you overlay on a
 * price chart (as lines/bands) or drop into a linked sub-pane (RSI, MACD).
 *
 * Every function returns a `Float64Array` the same length as its input, with a
 * leading run of `NaN` for the warm-up period (before enough data exists). Feed
 * the result straight to a line/area layer — plot helpers skip the NaN prefix.
 */

function toF64(values: ArrayLike<number>): Float64Array {
  const out = new Float64Array(values.length);
  for (let i = 0; i < values.length; i++) out[i] = values[i]!;
  return out;
}

/** Simple moving average over `period` samples. */
export function sma(values: ArrayLike<number>, period: number): Float64Array {
  const n = values.length;
  const out = new Float64Array(n).fill(NaN);
  if (period <= 0 || n < period) return out;
  let sum = 0;
  for (let i = 0; i < n; i++) {
    sum += values[i]!;
    if (i >= period) sum -= values[i - period]!;
    if (i >= period - 1) out[i] = sum / period;
  }
  return out;
}

/** Weighted moving average (linear weights 1..period, newest heaviest). */
export function wma(values: ArrayLike<number>, period: number): Float64Array {
  const n = values.length;
  const out = new Float64Array(n).fill(NaN);
  if (period <= 0 || n < period) return out;
  const denom = (period * (period + 1)) / 2;
  for (let i = period - 1; i < n; i++) {
    let acc = 0;
    for (let k = 0; k < period; k++) acc += values[i - period + 1 + k]! * (k + 1);
    out[i] = acc / denom;
  }
  return out;
}

/**
 * Exponential moving average, α = 2/(period+1). Seeded with the SMA of the first
 * `period` samples at index `period-1` (the standard convention).
 */
export function ema(values: ArrayLike<number>, period: number): Float64Array {
  const n = values.length;
  const out = new Float64Array(n).fill(NaN);
  if (period <= 0 || n < period) return out;
  const alpha = 2 / (period + 1);
  let seed = 0;
  for (let i = 0; i < period; i++) seed += values[i]!;
  let prev = seed / period;
  out[period - 1] = prev;
  for (let i = period; i < n; i++) {
    prev = values[i]! * alpha + prev * (1 - alpha);
    out[i] = prev;
  }
  return out;
}

/** Rolling population standard deviation over `period` samples. */
export function rollingStd(values: ArrayLike<number>, period: number): Float64Array {
  const n = values.length;
  const out = new Float64Array(n).fill(NaN);
  if (period <= 0 || n < period) return out;
  let sum = 0, sumSq = 0;
  for (let i = 0; i < n; i++) {
    const v = values[i]!;
    sum += v; sumSq += v * v;
    if (i >= period) { const old = values[i - period]!; sum -= old; sumSq -= old * old; }
    if (i >= period - 1) {
      const mean = sum / period;
      const variance = Math.max(0, sumSq / period - mean * mean);
      out[i] = Math.sqrt(variance);
    }
  }
  return out;
}

export interface BollingerBands {
  middle: Float64Array;
  upper: Float64Array;
  lower: Float64Array;
}

/** Bollinger Bands: SMA(period) ± `k`·rollingStd(period). Defaults 20, 2. */
export function bollinger(close: ArrayLike<number>, period = 20, k = 2): BollingerBands {
  const middle = sma(close, period);
  const sd = rollingStd(close, period);
  const n = close.length;
  const upper = new Float64Array(n).fill(NaN);
  const lower = new Float64Array(n).fill(NaN);
  for (let i = 0; i < n; i++) {
    if (!Number.isNaN(middle[i]!)) {
      upper[i] = middle[i]! + k * sd[i]!;
      lower[i] = middle[i]! - k * sd[i]!;
    }
  }
  return { middle, upper, lower };
}

/** Wilder's RSI over `period` (default 14). Values in 0..100; warm-up is NaN. */
export function rsi(close: ArrayLike<number>, period = 14): Float64Array {
  const n = close.length;
  const out = new Float64Array(n).fill(NaN);
  if (period <= 0 || n <= period) return out;
  let gain = 0, loss = 0;
  // Seed with the average gain/loss over the first `period` changes.
  for (let i = 1; i <= period; i++) {
    const ch = close[i]! - close[i - 1]!;
    if (ch >= 0) gain += ch; else loss -= ch;
  }
  let avgGain = gain / period, avgLoss = loss / period;
  out[period] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);
  for (let i = period + 1; i < n; i++) {
    const ch = close[i]! - close[i - 1]!;
    const g = ch > 0 ? ch : 0, l = ch < 0 ? -ch : 0;
    avgGain = (avgGain * (period - 1) + g) / period;
    avgLoss = (avgLoss * (period - 1) + l) / period;
    out[i] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);
  }
  return out;
}

export interface Macd {
  macd: Float64Array;
  signal: Float64Array;
  histogram: Float64Array;
}

/** MACD: EMA(fast) − EMA(slow), its EMA(signal) line, and the histogram. Defaults 12/26/9. */
export function macd(close: ArrayLike<number>, fast = 12, slow = 26, signalPeriod = 9): Macd {
  const n = close.length;
  const emaFast = ema(close, fast);
  const emaSlow = ema(close, slow);
  const line = new Float64Array(n).fill(NaN);
  for (let i = 0; i < n; i++) {
    if (!Number.isNaN(emaFast[i]!) && !Number.isNaN(emaSlow[i]!)) line[i] = emaFast[i]! - emaSlow[i]!;
  }
  // Signal = EMA of the MACD line over its valid (non-NaN) region.
  const firstValid = line.findIndex((v) => !Number.isNaN(v));
  const signal = new Float64Array(n).fill(NaN);
  if (firstValid >= 0) {
    const seg = line.subarray(firstValid);
    const sig = ema(seg, signalPeriod);
    for (let i = 0; i < sig.length; i++) signal[firstValid + i] = sig[i]!;
  }
  const histogram = new Float64Array(n).fill(NaN);
  for (let i = 0; i < n; i++) {
    if (!Number.isNaN(line[i]!) && !Number.isNaN(signal[i]!)) histogram[i] = line[i]! - signal[i]!;
  }
  return { macd: line, signal, histogram };
}

/**
 * Volume-weighted average price, cumulative from the first sample: running
 * Σ(typical·volume) / Σ(volume), where typical = (high+low+close)/3.
 */
export function vwap(
  high: ArrayLike<number>, low: ArrayLike<number>, close: ArrayLike<number>, volume: ArrayLike<number>,
): Float64Array {
  const n = Math.min(high.length, low.length, close.length, volume.length);
  const out = new Float64Array(n).fill(NaN);
  let cumPV = 0, cumV = 0;
  for (let i = 0; i < n; i++) {
    const typical = (high[i]! + low[i]! + close[i]!) / 3;
    cumPV += typical * volume[i]!;
    cumV += volume[i]!;
    out[i] = cumV === 0 ? NaN : cumPV / cumV;
  }
  return out;
}

/** True range per bar: max(H−L, |H−prevC|, |L−prevC|). First bar is H−L. */
export function trueRange(
  high: ArrayLike<number>, low: ArrayLike<number>, close: ArrayLike<number>,
): Float64Array {
  const n = Math.min(high.length, low.length, close.length);
  const out = new Float64Array(n);
  if (n > 0) out[0] = high[0]! - low[0]!;
  for (let i = 1; i < n; i++) {
    const pc = close[i - 1]!;
    out[i] = Math.max(high[i]! - low[i]!, Math.abs(high[i]! - pc), Math.abs(low[i]! - pc));
  }
  return out;
}

/** Wilder's Average True Range over `period` (default 14). */
export function atr(
  high: ArrayLike<number>, low: ArrayLike<number>, close: ArrayLike<number>, period = 14,
): Float64Array {
  const tr = trueRange(high, low, close);
  const n = tr.length;
  const out = new Float64Array(n).fill(NaN);
  if (n < period || period <= 0) return out;
  let sum = 0;
  for (let i = 0; i < period; i++) sum += tr[i]!;
  let prev = sum / period;
  out[period - 1] = prev;
  for (let i = period; i < n; i++) {
    prev = (prev * (period - 1) + tr[i]!) / period;
    out[i] = prev;
  }
  return out;
}

/** Index of the first non-NaN value (−1 if all NaN). Handy for trimming warm-up. */
export function firstFinite(values: ArrayLike<number>): number {
  for (let i = 0; i < values.length; i++) if (!Number.isNaN(values[i]!)) return i;
  return -1;
}

export { toF64 as toFloat64 };

File: native/src/finance/indicators.cpp
#include "finance/indicators.hpp"

#include <algorithm>
#include <cmath>
#include <limits>

namespace photon::finance {

namespace {

constexpr double kNaN = std::numeric_limits<double>::quiet_NaN();

Series warm(size_t n) { return Series(n, kNaN); }

/// Rolling extreme over a trailing window, NaN warm-up. O(n*period), which is
/// what the TypeScript does; a monotonic deque would be faster and would also
/// have to reproduce the >= / <= tie-breaks exactly, so it is not worth it at
/// the sizes an indicator runs on.
Series rolling_high(const double* v, size_t n, int period) {
  Series out = warm(n);
  if (period <= 0) return out;
  const size_t p = static_cast<size_t>(period);
  for (size_t i = p - 1; i < n; ++i) {
    double m = -std::numeric_limits<double>::infinity();
    for (size_t k = 0; k < p; ++k) m = std::max(m, v[i - k]);
    out[i] = m;
  }
  return out;
}

Series rolling_low(const double* v, size_t n, int period) {
  Series out = warm(n);
  if (period <= 0) return out;
  const size_t p = static_cast<size_t>(period);
  for (size_t i = p - 1; i < n; ++i) {
    double m = std::numeric_limits<double>::infinity();
    for (size_t k = 0; k < p; ++k) m = std::min(m, v[i - k]);
    out[i] = m;
  }
  return out;
}

}  // namespace

Series sma(const double* values, size_t count, int period) {
  Series out = warm(count);
  if (!values || period <= 0 || count < static_cast<size_t>(period)) return out;
  const size_t p = static_cast<size_t>(period);
  double sum = 0.0;
  for (size_t i = 0; i < count; ++i) {
    sum += values[i];
    if (i >= p) sum -= values[i - p];
    if (i >= p - 1) out[i] = sum / static_cast<double>(p);
  }
  return out;
}

Series wma(const double* values, size_t count, int period) {
  Series out = warm(count);
  if (!values || period <= 0 || count < static_cast<size_t>(period)) return out;
  const size_t p = static_cast<size_t>(period);
  const double denom = static_cast<double>(period * (period + 1)) / 2.0;
  for (size_t i = p - 1; i < count; ++i) {
    double acc = 0.0;
    for (size_t k = 0; k < p; ++k) acc += values[i - p + 1 + k] * static_cast<double>(k + 1);
    out[i] = acc / denom;
  }
  return out;
}

Series ema(const double* values, size_t count, int period) {
  Series out = warm(count);
  if (!values || period <= 0 || count < static_cast<size_t>(period)) return out;
  const size_t p = static_cast<size_t>(period);
  const double alpha = 2.0 / (static_cast<double>(period) + 1.0);
  double seed = 0.0;
  for (size_t i = 0; i < p; ++i) seed += values[i];
  double prev = seed / static_cast<double>(p);
  out[p - 1] = prev;
  for (size_t i = p; i < count; ++i) {
    prev = values[i] * alpha + prev * (1.0 - alpha);
    out[i] = prev;
  }
  return out;
}

Series rolling_std(const double* values, size_t count, int period) {
  Series out = warm(count);
  if (!values || period <= 0 || count < static_cast<size_t>(period)) return out;
  const size_t p = static_cast<size_t>(period);
  double sum = 0.0;
  double sum_sq = 0.0;
  for (size_t i = 0; i < count; ++i) {
    const double v = values[i];
    sum += v;
    sum_sq += v * v;
    if (i >= p) {
      const double old = values[i - p];
      sum -= old;
      sum_sq -= old * old;
    }
    if (i >= p - 1) {
      const double mean = sum / static_cast<double>(p);
      const double variance = std::max(0.0, sum_sq / static_cast<double>(p) - mean * mean);
      out[i] = std::sqrt(variance);
    }
  }
  return out;
}

Band bollinger(const double* close, size_t count, int period, double k) {
  Band out;
  out.middle = sma(close, count, period);
  const Series sd = rolling_std(close, count, period);
  out.upper = warm(count);
  out.lower = warm(count);
  for (size_t i = 0; i < count; ++i) {
    if (!std::isnan(out.middle[i])) {
      out.upper[i] = out.middle[i] + k * sd[i];
      out.lower[i] = out.middle[i] - k * sd[i];
    }
  }
  return out;
}

Series rsi(const double* close, size_t count, int period) {
  Series out = warm(count);
  if (!close || period <= 0 || count <= static_cast<size_t>(period)) return out;
  const size_t p = static_cast<size_t>(period);
  const double dp = static_cast<double>(period);
  double gain = 0.0;
  double loss = 0.0;
  for (size_t i = 1; i <= p; ++i) {
    const double ch = close[i] - close[i - 1];
    if (ch >= 0.0) {
      gain += ch;
    } else {
      loss -= ch;
    }
  }
  double avg_gain = gain / dp;
  double avg_loss = loss / dp;
  out[p] = avg_loss == 0.0 ? 100.0 : 100.0 - 100.0 / (1.0 + avg_gain / avg_loss);
  for (size_t i = p + 1; i < count; ++i) {
    const double ch = close[i] - close[i - 1];
    const double g = ch > 0.0 ? ch : 0.0;
    const double l = ch < 0.0 ? -ch : 0.0;
    avg_gain = (avg_gain * (dp - 1.0) + g) / dp;
    avg_loss = (avg_loss * (dp - 1.0) + l) / dp;
    out[i] = avg_loss == 0.0 ? 100.0 : 100.0 - 100.0 / (1.0 + avg_gain / avg_loss);
  }
  return out;
}

Macd macd(const double* close, size_t count, int fast, int slow, int signal_period) {
  Macd out;
  const Series ema_fast = ema(close, count, fast);
  const Series ema_slow = ema(close, count, slow);
  out.macd = warm(count);
  for (size_t i = 0; i < count; ++i) {
    if (!std::isnan(ema_fast[i]) && !std::isnan(ema_slow[i])) out.macd[i] = ema_fast[i] - ema_slow[i];
  }
  // The signal is the EMA of the MACD line over its *valid* region only: run it
  // across the NaN prefix and the seed would be NaN and stay that way forever.
  out.signal = warm(count);
  const int first = first_finite(out.macd.data(), count);
  if (first >= 0) {
    const size_t offset = static_cast<size_t>(first);
    const Series sig = ema(out.macd.data() + offset, count - offset, signal_period);
    for (size_t i = 0; i < sig.size(); ++i) out.signal[offset + i] = sig[i];
  }
  out.histogram = warm(count);
  for (size_t i = 0; i < count; ++i) {
    if (!std::isnan(out.macd[i]) && !std::isnan(out.signal[i])) {
      out.histogram[i] = out.macd[i] - out.signal[i];
    }
  }
  return out;
}

Series vwap(const double* high, const double* low, const double* close, const double* volume,
            size_t count) {
  Series out = warm(count);
  if (!high || !low || !close || !volume) return out;
  double cum_pv = 0.0;
  double cum_v = 0.0;
  for (size_t i = 0; i < count; ++i) {
    const double typical = (high[i] + low[i] + close[i]) / 3.0;
    cum_pv += typical * volume[i];
    cum_v += volume[i];
    out[i] = cum_v == 0.0 ? kNaN : cum_pv / cum_v;
  }
  return out;
}

Series true_range(const double* high, const double* low, const double* close, size_t count) {
  Series out(count, 0.0);
  if (!high || !low || !close || count == 0) return out;
  out[0] = high[0] - low[0];
  for (size_t i = 1; i < count; ++i) {
    const double pc = close[i - 1];
    out[i] = std::max({high[i] - low[i], std::abs(high[i] - pc), std::abs(low[i] - pc)});
  }
  return out;
}

Series atr(const double* high, const double* low, const double* close, size_t count, int period) {
  const Series tr = true_range(high, low, close, count);
  Series out = warm(count);
  if (period <= 0 || count < static_cast<size_t>(period)) return out;
  const size_t p = static_cast<size_t>(period);
  const double dp = static_cast<double>(period);
  double sum = 0.0;
  for (size_t i = 0; i < p; ++i) sum += tr[i];
  double prev = sum / dp;
  out[p - 1] = prev;
  for (size_t i = p; i < count; ++i) {
    prev = (prev * (dp - 1.0) + tr[i]) / dp;
    out[i] = prev;
  }
  return out;
}

int first_finite(const double* values, size_t count) {
  if (!values) return -1;
  for (size_t i = 0; i < count; ++i) {
    if (!std::isnan(values[i])) return static_cast<int>(i);
  }
  return -1;
}

Stochastic stochastic(const double* high, const double* low, const double* close, size_t count,
                      int k_period, int d_period) {
  Stochastic out;
  const Series hh = rolling_high(high, count, k_period);
  const Series ll = rolling_low(low, count, k_period);
  out.k = warm(count);
  for (size_t i = 0; i < count; ++i) {
    if (!std::isnan(hh[i])) {
      const double range = hh[i] - ll[i];
      out.k[i] = range == 0.0 ? 50.0 : (100.0 * (close[i] - ll[i])) / range;
    }
  }
  out.d = sma(out.k.data(), count, d_period);
  return out;
}

Band keltner(const double* high, const double* low, const double* close, size_t count, int period,
             double mult, int atr_period) {
  Band out;
  out.middle = ema(close, count, period);
  const Series a = atr(high, low, close, count, atr_period);
  out.upper = warm(count);
  out.lower = warm(count);
  for (size_t i = 0; i < count; ++i) {
    if (!std::isnan(out.middle[i]) && !std::isnan(a[i])) {
      out.upper[i] = out.middle[i] + mult * a[i];
      out.lower[i] = out.middle[i] - mult * a[i];
    }
  }
  return out;
}

Series obv(const double* close, const double* volume, size_t count) {
  Series out(count, 0.0);
  if (!close || !volume) return out;
  for (size_t i = 1; i < count; ++i) {
    const double d = close[i] - close[i - 1];
    out[i] = out[i - 1] + (d > 0.0 ? volume[i] : d < 0.0 ? -volume[i] : 0.0);
  }
  return out;
}

Ichimoku ichimoku(const double* high, const double* low, size_t count, int conv_period,
                  int base_period, int span_b_period) {
  const auto mid = [&](int p) {
    const Series hi = rolling_high(high, count, p);
    const Series lo = rolling_low(low, count, p);
    Series o = warm(count);
    for (size_t i = 0; i < count; ++i) {
      if (!std::isnan(hi[i])) o[i] = (hi[i] + lo[i]) / 2.0;
    }
    return o;
  };
  Ichimoku out;
  out.conversion = mid(conv_period);
  out.base = mid(base_period);
  out.span_b = mid(span_b_period);
  out.span_a = warm(count);
  for (size_t i = 0; i < count; ++i) {
    if (!std::isnan(out.conversion[i]) && !std::isnan(out.base[i])) {
      out.span_a[i] = (out.conversion[i] + out.base[i]) / 2.0;
    }
  }
  return out;
}

Adx adx(const double* high, const double* low, const double* close, size_t count, int period) {
  Adx out;
  out.adx = warm(count);
  out.plus_di = warm(count);
  out.minus_di = warm(count);
  if (!high || !low || !close || period <= 0 || count <= static_cast<size_t>(period)) return out;
  const size_t p = static_cast<size_t>(period);
  const double dp = static_cast<double>(period);

  Series tr(count, 0.0);
  Series pdm(count, 0.0);
  Series mdm(count, 0.0);
  for (size_t i = 1; i < count; ++i) {
    const double up = high[i] - high[i - 1];
    const double down = low[i - 1] - low[i];
    pdm[i] = (up > down && up > 0.0) ? up : 0.0;
    mdm[i] = (down > up && down > 0.0) ? down : 0.0;
    const double pc = close[i - 1];
    tr[i] = std::max({high[i] - low[i], std::abs(high[i] - pc), std::abs(low[i] - pc)});
  }

  // Wilder-smoothed sums, seeded over the first `period` changes.
  double s_tr = 0.0;
  double s_p = 0.0;
  double s_m = 0.0;
  for (size_t i = 1; i <= p; ++i) {
    s_tr += tr[i];
    s_p += pdm[i];
    s_m += mdm[i];
  }
  Series dx = warm(count);
  const auto di_at = [&](size_t i) {
    const double plus = s_tr == 0.0 ? 0.0 : (100.0 * s_p) / s_tr;
    const double minus = s_tr == 0.0 ? 0.0 : (100.0 * s_m) / s_tr;
    out.plus_di[i] = plus;
    out.minus_di[i] = minus;
    const double sum = plus + minus;
    dx[i] = sum == 0.0 ? 0.0 : (100.0 * std::abs(plus - minus)) / sum;
  };
  di_at(p);
  for (size_t i = p + 1; i < count; ++i) {
    s_tr = s_tr - s_tr / dp + tr[i];
    s_p = s_p - s_p / dp + pdm[i];
    s_m = s_m - s_m / dp + mdm[i];
    di_at(i);
  }

  // ADX is the Wilder-smoothed DX, starting `period` bars after the first DX.
  const size_t adx_start = p + p;
  if (adx_start < count) {
    double acc = 0.0;
    for (size_t i = p + 1; i <= adx_start; ++i) acc += dx[i];
    double prev = acc / dp;
    out.adx[adx_start] = prev;
    for (size_t i = adx_start + 1; i < count; ++i) {
      prev = (prev * (dp - 1.0) + dx[i]) / dp;
      out.adx[i] = prev;
    }
  }
  return out;
}

SuperTrend super_trend(const double* high, const double* low, const double* close, size_t count,
                       int period, double mult) {
  SuperTrend out;
  out.trend = warm(count);
  out.direction = warm(count);
  if (!high || !low || !close) return out;
  const Series a = atr(high, low, close, count, period);
  Series final_upper(count, 0.0);
  Series final_lower(count, 0.0);
  int dir = 1;
  for (size_t i = 0; i < count; ++i) {
    if (std::isnan(a[i])) continue;
    const double hl2 = (high[i] + low[i]) / 2.0;
    const double basic_upper = hl2 + mult * a[i];
    const double basic_lower = hl2 - mult * a[i];
    const bool prev_valid = i > 0 && !std::isnan(a[i - 1]);
    final_upper[i] =
        prev_valid && (basic_upper < final_upper[i - 1] || close[i - 1] > final_upper[i - 1])
            ? basic_upper
            : (prev_valid ? final_upper[i - 1] : basic_upper);
    final_lower[i] =
        prev_valid && (basic_lower > final_lower[i - 1] || close[i - 1] < final_lower[i - 1])
            ? basic_lower
            : (prev_valid ? final_lower[i - 1] : basic_lower);
    if (!prev_valid) {
      dir = close[i] >= hl2 ? 1 : -1;
    } else if (dir == 1 && close[i] < final_lower[i]) {
      dir = -1;
    } else if (dir == -1 && close[i] > final_upper[i]) {
      dir = 1;
    }
    out.direction[i] = dir;
    out.trend[i] = dir == 1 ? final_lower[i] : final_upper[i];
  }
  return out;
}

Series cci(const double* high, const double* low, const double* close, size_t count, int period) {
  Series out = warm(count);
  if (!high || !low || !close || period <= 0 || count < static_cast<size_t>(period)) return out;
  const size_t p = static_cast<size_t>(period);
  Series tp(count, 0.0);
  for (size_t i = 0; i < count; ++i) tp[i] = (high[i] + low[i] + close[i]) / 3.0;
  const Series avg = sma(tp.data(), count, period);
  for (size_t i = p - 1; i < count; ++i) {
    double dev = 0.0;
    for (size_t k = i - p + 1; k <= i; ++k) dev += std::abs(tp[k] - avg[i]);
    dev /= static_cast<double>(p);
    out[i] = dev == 0.0 ? 0.0 : (tp[i] - avg[i]) / (0.015 * dev);
  }
  return out;
}

Series mfi(const double* high, const double* low, const double* close, const double* volume,
           size_t count, int period) {
  Series out = warm(count);
  if (!high || !low || !close || !volume || period <= 0 || count <= static_cast<size_t>(period)) {
    return out;
  }
  const size_t p = static_cast<size_t>(period);
  Series tp(count, 0.0);
  for (size_t i = 0; i < count; ++i) tp[i] = (high[i] + low[i] + close[i]) / 3.0;
  for (size_t i = p; i < count; ++i) {
    double pos = 0.0;
    double neg = 0.0;
    for (size_t k = i - p + 1; k <= i; ++k) {
      const double flow = tp[k] * volume[k];
      if (tp[k] > tp[k - 1]) {
        pos += flow;
      } else if (tp[k] < tp[k - 1]) {
        neg += flow;
      }
    }
    out[i] = neg == 0.0 ? 100.0 : 100.0 - 100.0 / (1.0 + pos / neg);
  }
  return out;
}

Series williams_r(const double* high, const double* low, const double* close, size_t count,
                  int period) {
  Series out = warm(count);
  if (!high || !low || !close || period <= 0 || count < static_cast<size_t>(period)) return out;
  const size_t p = static_cast<size_t>(period);
  for (size_t i = p - 1; i < count; ++i) {
    double hh = -std::numeric_limits<double>::infinity();
    double ll = std::numeric_limits<double>::infinity();
    for (size_t k = i - p + 1; k <= i; ++k) {
      if (high[k] > hh) hh = high[k];
      if (low[k] < ll) ll = low[k];
    }
    const double span = hh - ll;
    out[i] = span == 0.0 ? 0.0 : ((hh - close[i]) / span) * -100.0;
  }
  return out;
}

Aroon aroon(const double* high, const double* low, size_t count, int period) {
  Aroon out;
  out.up = warm(count);
  out.down = warm(count);
  out.oscillator = warm(count);
  if (!high || !low || period <= 0 || count <= static_cast<size_t>(period)) return out;
  const size_t p = static_cast<size_t>(period);
  const double dp = static_cast<double>(period);
  for (size_t i = p; i < count; ++i) {
    double hi = -std::numeric_limits<double>::infinity();
    double lo = std::numeric_limits<double>::infinity();
    size_t h_idx = i;
    size_t l_idx = i;
    for (size_t k = i - p; k <= i; ++k) {
      if (high[k] >= hi) {
        hi = high[k];
        h_idx = k;
      }
      if (low[k] <= lo) {
        lo = low[k];
        l_idx = k;
      }
    }
    out.up[i] = ((dp - static_cast<double>(i - h_idx)) / dp) * 100.0;
    out.down[i] = ((dp - static_cast<double>(i - l_idx)) / dp) * 100.0;
    out.oscillator[i] = out.up[i] - out.down[i];
  }
  return out;
}

Band donchian(const double* high, const double* low, size_t count, int period) {
  Band out;
  out.middle = warm(count);
  out.upper = warm(count);
  out.lower = warm(count);
  if (!high || !low || period <= 0 || count < static_cast<size_t>(period)) return out;
  const size_t p = static_cast<size_t>(period);
  for (size_t i = p - 1; i < count; ++i) {
    double hh = -std::numeric_limits<double>::infinity();
    double ll = std::numeric_limits<double>::infinity();
    for (size_t k = i - p + 1; k <= i; ++k) {
      if (high[k] > hh) hh = high[k];
      if (low[k] < ll) ll = low[k];
    }
    out.upper[i] = hh;
    out.lower[i] = ll;
    out.middle[i] = (hh + ll) / 2.0;
  }
  return out;
}

Series parabolic_sar(const double* high, const double* low, size_t count, double step,
                     double max_step) {
  Series out = warm(count);
  if (!high || !low || count < 2) return out;
  bool rising = high[1] >= high[0];
  double sar = rising ? low[0] : high[0];
  double extreme = rising ? high[1] : low[1];
  double accel = step;
  out[1] = sar;
  for (size_t i = 2; i < count; ++i) {
    sar += accel * (extreme - sar);
    // The stop may never enter the previous two bars' range.
    if (rising) {
      sar = std::min({sar, low[i - 1], low[i - 2]});
    } else {
      sar = std::max({sar, high[i - 1], high[i - 2]});
    }
    if (rising ? low[i] < sar : high[i] > sar) {
      // Reverse: the stop becomes the extreme and a fresh trend starts.
      rising = !rising;
      sar = extreme;
      extreme = rising ? high[i] : low[i];
      accel = step;
    } else if (rising ? high[i] > extreme : low[i] < extreme) {
      extreme = rising ? high[i] : low[i];
      accel = std::min(max_step, accel + step);
    }
    out[i] = sar;
  }
  return out;
}

PivotLevels pivot_points(double high, double low, double close) {
  const double pivot = (high + low + close) / 3.0;
  const double range = high - low;
  PivotLevels out{};
  out.pivot = pivot;
  out.r1 = 2.0 * pivot - low;
  out.s1 = 2.0 * pivot - high;
  out.r2 = pivot + range;
  out.s2 = pivot - range;
  out.r3 = high + 2.0 * (pivot - low);
  out.s3 = low - 2.0 * (high - pivot);
  return out;
}

}  // namespace photon::finance
